### DISCUSSION TITLE
Viability of Deck.gl due to Mapbox GL licensing change

### DISCUSSION THREAD

**Discussion by ThomasG77** [visgl/deck.gl#5237] [General]
See announcement https://github.com/mapbox/mapbox-gl-js/releases/tag/v2.0.0 and associated discussion
https://news.ycombinator.com/item?id=25347310
What does it mean for Deck.GL future?

**Comment by Pessimistress** [comment]
First, let's make it very clear: deck.gl is not dependent on mapbox-gl. We used Mapbox as a base map provider in many of our demos because it was open source and flexible. However, in the end it's just one of the base map providers that we support, no different from Google Maps and ArcGIS.

Now regarding this change: from what I understand, Mapbox GL JS is going "commercial source" instead of "closed source." For the real-world developers who are already using Mapbox as a paid product, it doesn't make a difference. For those who don't, there are alternatives to choose from, justifiably with some annoyance. For the deck.gl project, this certainly places some obstacles in the experience that we are shooting for, e.g. new developers may not be able to run our website examples in the future without a Mapbox account. When that happens, we will likely move to an alternative provider.

**Reply by ThomasG77** [reply]
Thanks for your feedback

**Reply by chrisgervang** [reply]
An open question I have is what are we allowed to do with their commercial source? For example, we currently mimic mapbox's camera settings so that they match across many attributes. We also have a `@deck.gl/mapbox` sub-module for weaving layers together between the libraries, which is awesome.

While these integrations haven't necessarily broken overnight, I'm curious if we're allowed to read their code to continue enhancing the integration between deck.gl and mapbox. For example, while I imagine the new terrain and sky layer work out of the box since they are just new layers, they've added a FreeCamera API that may not "just work" due to how the view model is manipulated. I'm curious how we can go about building a new compatible view that matches theirs?

**Comment by ibgreen** [comment]
Great questions. However given the concerns that users have around the mapbox licensing change, I will restate for the record that these questions have nothing to do with mapbox licenses. deck.gl does not fork, copy or depend on mapbox code.

Perhaps this question could be reopened as a separate topic: `How can deck.gl integrate with new features in mapbox v2?`

Quick answers:
- `FreeCamera`- deck.gl has its own implementation of viewport parameters. It is true that we will need to match mapbox free camera parameters, but we don't need to fork any code to do that.
- `@deck.glmapbox` is built on public mapbox APIs and should continue to work as long as mapbox supports those (for 2D basemaps at least)

**Comment by ibgreen** [comment]
Regarding `visualization-on-top-of-3D-mapbox-terrain`
- Unfortunately this integration would likely be technically complicated, e.g. as a lot of compositing clearly happens in off-screen framebuffers which are not exposed via any mapbox APIs (such as the public mapbox APIs used by the `@deck.gl/mapbox` module).
- However, if mapbox does preserve the Z buffer in their final render buffer there may be some hope, at least for extruded/3D deck.gl layers.

I would personally recommend implementing 3D adjusted visualizations in deck itself first, before trying to align with mapbox terrain. 
- This feature was listed as a goal for deck.gl 8.5 in the last UCF open governance planning meeting, so the stage is set for contribution and collaboration. 
- The natural initial integration would be with the internal deck.gl `TerrainLayer` as a basemap.
